contract_invoicing_grouping_criteria/models/contract.py

```diff
@@ -0,0 +1,188 @@
+from itertools import groupby
+
+from odoo import models
+
+
+def group_on_keys(keys, get_ids=False):
+    """Return a grouping function on the given keys"""
+
+    def id_if_is_record(x):
+        # if x is a record, return its id to get a consistent sorting
+        return x.id if isinstance(x, models.BaseModel) else x
+
+    def get_with_or(x, key):
+        """Get key from x. Handles keys containing ||"""
+        for k in key.split("||"):
+            val = getattr(x, k)
+            if val:
+                return val
+        return val
```

Comment:
val must be defined before the for loop if key is empty we can have a bug.

Reply:
Split with an argument always returns a non empty list
(cf: https://docs.python.org/3.3/library/stdtypes.html#str.split)

Reply:
in this case the last return is useless ?

Reply:
Not very useful indeed but this does not matter much